### PR TITLE
webhook-apicoverage: HTML display and other fixes

### PR DESCRIPTION
**Pull request by dushyanthsc**
Based on the feedback for the tool, this change addresses:

  1. Wrapping JSON like display inside a HTML page and color code covered(green) and notcovered(red) 
      fields for better readability.
  2. Fixes the total field calculation by taking into account ignored fields.
  3. Ensures we visit a node type only once and prune the traversal if a same node type appears again in 
      other parts of resource tree.

### CODE REVIEW

**Comment by dushyanthsc** [comment]
/hold

@yt3liu  pointed me to Go HTML template support, which is a cleaner way to generate the HTML page, I am evaluating and if suitable applying the approach to this change.

**Comment by dushyanthsc** [comment]
/hold cancel

I took a look at the Go's template support and it is the right way to do this. However handling loops and if constructs is a bit complicated. Considering other important items in this space, I am going to punt on this but have created Issue: https://github.com/knative/test-infra/issues/577 to make sure I fix this.

File: tools/webhook-apicoverage/view/html_display.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/knative/test-infra/tools/webhook-apicoverage/coveragecalculator"
+)
+
+// HTMLHeader sets up an HTML page with the right style format
+var HTMLHeader = fmt.Sprintf(`
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/view/html_display.go]: you dont need fmt.Sprintf here you can have 
var a:=`
foob
bar`

File: tools/webhook-apicoverage/view/html_display.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+table, th, td { border: 1px solid white; text-align: center}
+
+.braces {color: white; size: A3}
+-->
+</style>
+<body style="background-color:rgb(0,0,0); font-family: Arial">
+`)
+
+// HTMLFooter closes the tags for the HTML page.
+var HTMLFooter = fmt.Sprintf(`
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/view/html_display.go]: ditto

File: tools/webhook-apicoverage/view/html_display.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+		}
+		buffer.WriteString(fmt.Sprint(`<div class="styleheader">`))
+		buffer.WriteString(fmt.Sprintf(`<br>Package: %s`, packageName))
+		buffer.WriteString(fmt.Sprintf(`<br>Type: %s`, typeName))
+		buffer.WriteString(fmt.Sprint(`<br></div>`))
+		buffer.WriteString(fmt.Sprint(`<div class="braces"><br>{</div>`))
+		for _, fieldCoverage := range typeCoverage.Fields {
+			var fieldDisplay string
+			if displayRules.FieldRule != nil {
+				fieldDisplay = displayRules.FieldRule(fieldCoverage)
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/view/html_display.go]: no need of additional var.
if(!nil) {
WriteString(displayRules.FieldRule(fieldCoverage))
} else {
WriteString(defaultHTMLTypeDisplay(fieldCoverage))
}

File: tools/webhook-apicoverage/view/html_display.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+			} else {
+				fieldDisplay = defaultHTMLTypeDisplay(fieldCoverage)
+			}
+			buffer.WriteString(fieldDisplay)
+		}
+
+		buffer.WriteString(fmt.Sprint(`<div class="braces">}</div>`))
+		buffer.WriteString(fmt.Sprint(`<br>`))
+
```
Comment (srinivashegde86) [inline tools/webhook-apicoverage/view/html_display.go]: remove extra line

**Comment by dushyanthsc** [comment]
Some of the items are going to cleared with HTML improvement issue, I will club the others in that change itself.

File: tools/webhook-apicoverage/view/html_display.go (hunk 1 above)
Comment (adrcunha) [inline tools/webhook-apicoverage/view/html_display.go]: nit: make this a const.

File: tools/webhook-apicoverage/view/html_display.go (hunk 2 above)
Comment (adrcunha) [inline tools/webhook-apicoverage/view/html_display.go]: Ditto.

File: tools/webhook-apicoverage/view/html_display.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+.ignored {color: white; size: A4}
+
+.values {color: yellow; size: A3}
+
+table, th, td { border: 1px solid white; text-align: center}
+
+.braces {color: white; size: A3}
+-->
+</style>
+<body style="background-color:rgb(0,0,0); font-family: Arial">
```
Comment (adrcunha) [inline tools/webhook-apicoverage/view/html_display.go]: Would Courier/monospaced be more suitable for the JSON dump?

**Pull request merged**